plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1167 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "kafka.hpp"
+
+#include "rtlds_imp.hpp"
+#include "jlog.hpp"
+#include "jmutex.hpp"
+#include "jprop.hpp"
+#include "jfile.hpp"
+#include "build-config.h"
+
+#include "librdkafka/rdkafka.h"
+
+#include <map>
+#include <fstream>
+
+//==============================================================================
+// Kafka Interface Code
+//==============================================================================
+
+namespace KafkaPlugin
+{
+    //--------------------------------------------------------------------------
+    // File Constants
+    //--------------------------------------------------------------------------
+
+    // The minimum number of seconds that a cached object can live
+    // without activity
+    const time_t    OBJECT_EXPIRE_TIMEOUT_SECONDS = 60 * 2;
+
+    // The number of milliseconds given to librdkafka to perform explicit
+    // background activity
+    const __int32   POLL_TIMEOUT = 1000;
+
+    //--------------------------------------------------------------------------
+    // Static Variables
+    //--------------------------------------------------------------------------
+
+    static Mutex    initLock;
+
+    //--------------------------------------------------------------------------
+    // Static Methods (internal)
+    //--------------------------------------------------------------------------
+
+    /**
+     * Look for an optional configuration file and apply any found configuration
+     * parameters to a librdkafka configuration object.
+     *
+     * @param   configFilePath      The path to a configuration file; it is not
+     *                              necessary for the file to exist
+     * @param   globalConfigPtr     A pointer to the configuration object that
+     *                              will receive any found parameters
+     */
+    static void applyConfig(const char* configFilePath, RdKafka::Conf* globalConfigPtr)
+    {
+        if (configFilePath && *configFilePath && globalConfigPtr)
+        {
+            std::string                 errStr;
+            StringBuffer                fullConfigPath;
+
+            fullConfigPath.append(CONFIG_DIR).append(PATHSEPSTR).append(configFilePath);
+
+            Owned<IProperties>          properties = createProperties(fullConfigPath.str(), true);
+            Owned<IPropertyIterator>    props = properties->getIterator();
+
+            ForEach(*props)
+            {
+                StringBuffer key = props->getPropKey();
+
+                if (key.length() > 0 && key.charAt(0) != '#')
+                {
+                    if (strcmp(key.str(), "metadata.broker.list") != 0)
+                    {
+                        const char*     value = properties->queryProp(key);
+
```

Comment:
A (very) few empty lines were a mistake.  Most are simply style, to vertically segment code for theoretical improved readability.  I can make the code more compact if you prefer.